docs/design/2021-03-30-small-table-caching.md

```diff
@@ -0,0 +1,208 @@
+
+When performing a modification, the meta-information should be checked first. If there is a "read" lock, the “read” lock needs to be upgraded to a "write" lock. This step prevents the read lock from renewing the lease. After the lease expires, the write operation can be performed. And after the write operation finishes, the "write" lock should be cleared, so that the subsequent "read" lock can be added and the data can be cached again.
+
+The "write" lock also needs a TTL(time to live) to handle abnormal situations. For example, if the  TiDB server crashes after the write lock succeeds, TTL gives us a hint of how to clean the orphan write lock.
+
+To put all things together, we have this interface:
+
+```
+type StateRemote interface {
+    LockForRead()
+    RenewLease()
+    LockForWrite()
+    WriteAndUnlock()
+    CleanOrphanLock()
+}
+```
+
+### Management of lock meta information
+
+What’s the granularity of the lock? One option is to implement the meta lock at the key-value level. The advantage is that the granularity is finer and the updating performance is better. If we cache the key-value record, modifying one key would not affect the cache of the other keys.  The disadvantage is that this will invade the current transaction layer implementation, and make our transaction more complex. Considering the current implementation of the transaction is very complex already, it is not wise to do so.
+
+The other option is to implement the meta lock at the table level. Each table corresponds to one meta lock record. As long as any key of the table is modified, the table cache becomes invalid. Assuming that we only deal with small tables, and this table is updated infrequently, this option should be acceptable.
+
+For updating operation, the `StateRemote` interface needs to be implemented. The following part in this section describes the management of meta-information, where it should be stored, and how it is used.
+
+Such a table could be used to store the meta information:
+
+```
+CREATE TABLE mysql.table_cache_meta (
+   tid int primary key,
+   lock_type enum('NONE','READ','WRITE'),
+   lease uint64,
+);
+```
+
+This table contains the `tid`, `lock_type` and `lease` information.
+- `tid` records the ID of the table being cached
+- `lock_type` decides whether the current state is read locked or write locked,
+- `lease` is used for the read locks lease renewal, and also for cleanup the write locks
+
+LockForRead() corresponds to this operation:
+
+```
+update table_cache_meta set lock ='READ', lease = %? where  tid = %? and lock !='WRITE'
+```
+
+RenewLease() corresponds to this operation:
+
+```
+update table_cache_meta set lease = %? where tid = %? and lock ='READ'
+```
+
+LockForWrite() corresponds to this operation:
+
+```
+update table_cache_meta set lock ='WRITE', lease = %?' where tid = %?
```

Comment:
1. write needs to wait for read lease gone, so if sleep() in the code before this call

2. no starvation for write, because when the write set the state to Write Intend, the read lock can't renew its lease any more

Reply:
I'll update the document to clarify that